utility/session.go

```diff
@@ -1,157 +1,266 @@
 package utility
 
-// IMPORTANT: The interface and implementation defined in this file are for illustrative purposes only
-// and need to be revisited before any implementation commences.
-
 import (
 	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"math/rand"
 
+	"github.com/pokt-network/pocket/logger"
 	coreTypes "github.com/pokt-network/pocket/shared/core/types"
 	"github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/modules"
 	"github.com/pokt-network/pocket/utility/types"
+	"golang.org/x/exp/slices"
 )
 
-// TODO: When implementing please review if block height tolerance (+,-1) is included in the session protocol: pokt-network/pocket-core#1464 CC @Olshansk
-
-// REFACTOR: Move these into `utility/types` and consider creating an enum
-type RelayChain string
-type GeoZone string
-
-type Session interface {
-	NewSession(sessionHeight int64, relayChain RelayChain, geoZone GeoZone, application *coreTypes.Actor) (Session, types.Error)
-	GetSessionID() []byte             // the identifier of the dispatched session
-	GetSessionHeight() int64          // the block height when the session started
-	GetRelayChain() RelayChain        // the web3 chain identifier
-	GetGeoZone() GeoZone              // the geo-location zone where the application is intending to operate during the session
-	GetApplication() *coreTypes.Actor // the Application consuming the web3 access
-	GetServicers() []*coreTypes.Actor // the Servicers providing Web3 to the application
-	GetFishermen() []*coreTypes.Actor // the Fishermen monitoring the servicers
-}
+// GetSession implements of the exposed `UtilityModule.GetSession` function
+func (m *utilityModule) GetSession(appAddr string, height int64, relayChain, geoZone string) (*coreTypes.Session, error) {
+	persistenceModule := m.GetBus().GetPersistenceModule()
+	readCtx, err := persistenceModule.NewReadContext(height)
+	if err != nil {
+		return nil, err
+	}
+	defer readCtx.Release()
 
-var _ Session = &session{}
+	session := &coreTypes.Session{
+		RelayChain: relayChain,
+		GeoZone:    geoZone,
+	}
 
-type session struct {
-	sessionId   []byte
-	height      int64
-	relayChain  RelayChain
-	geoZone     GeoZone
-	application *coreTypes.Actor
-	servicers   []*coreTypes.Actor
-	fishermen   []*coreTypes.Actor
-}
+	sessionHydrator := &sessionHydrator{
+		logger:      m.logger.With().Str("source", "sessionHydrator").Logger(),
```

Comment:
Nice! I didn't know this was a thing. :raised_hands:

Reply:
I updated the logger README to reflect this.